haystack/components/readers/extractive.py

```diff
@@ -299,21 +308,122 @@ def _nest_answers(
                 answer.query = queries[query_id]
                 current_answers.append(answer)
                 i += 1
-            current_answers = sorted(current_answers, key=lambda answer: answer.score, reverse=True)
+            current_answers = sorted(current_answers, key=lambda ans: ans.score, reverse=True)
+            current_answers = self.deduplicate_by_overlap(current_answers, overlap_threshold=overlap_threshold)
             current_answers = current_answers[:top_k]
             if no_answer:
                 no_answer_score = math.prod(1 - answer.score for answer in current_answers)
                 answer_ = ExtractedAnswer(
                     data=None, query=queries[query_id], meta={}, document=None, score=no_answer_score
                 )
                 current_answers.append(answer_)
-            current_answers = sorted(current_answers, key=lambda answer: answer.score, reverse=True)
+            current_answers = sorted(current_answers, key=lambda ans: ans.score, reverse=True)
             if score_threshold is not None:
                 current_answers = [answer for answer in current_answers if answer.score >= score_threshold]
             nested_answers.append(current_answers)
 
         return nested_answers
 
+    def _calculate_overlap(self, answer1_start: int, answer1_end: int, answer2_start: int, answer2_end: int) -> int:
+        """
+        Calculates the amount of overlap (in number of characters) between two answer offsets.
+
+        Stack overflow post explaining how to calculate overlap between two ranges:
+            https://stackoverflow.com/questions/325933/determine-whether-two-date-ranges-overlap/325964#325964
+        """
+        start1, end1 = answer1_start, answer1_end
+        start2, end2 = answer2_start, answer2_end
```

Comment:
Why not just using the methods parameters?

Reply:
No good reason, I think this was a leftover when the input was a Span instead of 4 numbers. I'll go ahead and change